### PR TITLE
Android 13 support *unfinished*

### PR DESCRIPTION
**Pull request by amritk**
ref #218 

Just need a little bit of help with the final part of showing the dialog

### CODE REVIEW

File: src/android/FirebaseMessagingPlugin.java [hunk 1]
```diff
@@ -137,7 +142,14 @@ private void requestPermission(CordovaArgs args, CallbackContext callbackContext
         if (NotificationManagerCompat.from(context).areNotificationsEnabled()) {
             callbackContext.success();
         } else {
-            callbackContext.error("Notifications permission is not granted");
+            if (Build.VERSION.SDK_INT >= 32) {
```
Comment (chemerisuk) [inline src/android/FirebaseMessagingPlugin.java:145]: Should be 33
```suggestion
            if (Build.VERSION.SDK_INT >= 33) {
```

File: src/android/FirebaseMessagingPlugin.java [hunk 2]
```diff
@@ -137,7 +142,14 @@ private void requestPermission(CordovaArgs args, CallbackContext callbackContext
...
+                if (ContextCompat.checkSelfPermission(null,
```
Comment (chemerisuk) [inline src/android/FirebaseMessagingPlugin.java:146]: Use `cordova.hasPermission(Manifest.permission.POST_NOTIFICATIONS)` to verify if permission is already granted for cordova framework

File: src/android/FirebaseMessagingPlugin.java [hunk 3]
```diff
@@ -44,6 +47,8 @@ public class FirebaseMessagingPlugin extends ReflectiveCordovaPlugin {
     private NotificationManager notificationManager;
     private FirebaseMessaging firebaseMessaging;
 
+    private FirebaseMessagingActivity firebaseMessagingActivity = new FirebaseMessagingActivity();
```
Comment (chemerisuk) [inline src/android/FirebaseMessagingPlugin.java:50]: Better to keep launcher inline. You need to store private callbackContext in order to call it after getting feedback from user
```java
private final ActivityResultLauncher<String> requestPermissionLauncher =
        registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
            if (isGranted) {
                // FCM SDK (and your app) can post notifications.
                requiestPermissionCallback.success();
            } else {
                requiestPermissionCallback.error("Notifications permission is not granted");
            }
        });
```

File: src/android/FirebaseMessagingPlugin.java [hunk 4]
```diff
@@ -137,7 +142,14 @@ private void requestPermission(CordovaArgs args, CallbackContext callbackContext
         if (NotificationManagerCompat.from(context).areNotificationsEnabled()) {
             callbackContext.success();
         } else {
-            callbackContext.error("Notifications permission is not granted");
+            if (Build.VERSION.SDK_INT >= 32) {
+                if (ContextCompat.checkSelfPermission(null,
+                        Manifest.permission.ACCESS_NOTIFICATION_POLICY) == PackageManager.PERMISSION_GRANTED)
+                    return;
+                firebaseMessagingActivity.launchPermissions();
```
Comment (chemerisuk) [inline src/android/FirebaseMessagingPlugin.java:149]: Use private permission launcher and store `calbackContext ` for later usage
```suggestion
                requiestPermissionCallback = calbackContext;
                requestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS);
```

**Review by chemerisuk — CHANGES_REQUESTED** [review]
Hey @amritk, thanks for starting this PR. Android 13 is relatively new that's why I was planning to work on this later. Anyway there is a guide from official docs https://firebase.google.com/docs/cloud-messaging/android/client#request-permission13. Most of my comments below follow the guide

File: src/android/FirebaseMessagingPlugin.java (hunk 3 above)
Comment (amritk) [reply src/android/FirebaseMessagingPlugin.java:50]: Yea I was just trying to get it working first, then was going to propogate the callback

**Comment by amritk** [comment]
hey @chemerisuk thanks for the feedback! This was my first time writing anything native 😓 I will implement the changes.

File: src/android/FirebaseMessagingPlugin.java (hunk 3 above)
Comment (amritk) [reply src/android/FirebaseMessagingPlugin.java]: @chemerisuk So this is why I had that extra activity before, if we do it inline then 
`registerForActivityResult` cannot be resolved. What do I need to import to get it to work?

**Comment by amritk** [comment]
So now it builds, but it crashes the app

**Comment by chemerisuk** [comment]
LGTM, thanks @amritk !

**Comment by fluxc** [comment]
I have been testing this improvement and looks like is crashing the app. This is a log from adb logcat
image

**Comment by amritk** [comment]
Yea that was actually my last comment 

https://github.com/chemerisuk/cordova-plugin-firebase-messaging/pull/220#issuecomment-1317787083

@chemerisuk

**Comment by fluxc** [comment]
I manage to fix the issue here on this PR
https://github.com/chemerisuk/cordova-plugin-firebase-messaging/pull/227

**Comment by soerenBoisen** [comment]
@amritk @chemerisuk 
I think it's because you are trying to access cordova object too early:
https://stackoverflow.com/questions/25542790/why-cant-i-access-an-instance-of-cordovainterface-in-a-cordovaplugin-constructo

Would it make sense to initialize the `requestPermissionLauncher` from `pluginInitialize` method? That should solve the crash.

**Comment by tudordumitriu** [comment]
Hi guys
Is there a resolution / stable version with this fix? 
Since now Google Play has pushed the min target sdk to 33, if someone uses this plugin, the notifications still work, but the user has to go in manually to allow them.
So any ideas / help much appreciated!

**Pull request merged**